s3-streamer

```diff
@@ -0,0 +1,319 @@
+
+    def has(self, filename):
+        return os.path.exists(self.path(filename))
+
+    def write(self, filename, data):
+        print(f'Write {self.path(filename)}')
+        with open(self.path(filename), 'wb+') as file:
+            file.write(data)
+
+    def delete(self, filenames):
+        for filename in filenames:
+            print(f'Delete {self.path(filename)}')
+            os.unlink(self.path(filename))
+
+
+class S3Destination(Destination):
+    def __init__(self, directory):
+        self.directory = directory.rstrip('/') + '/'
+
+    def url(self, filename):
+        return urllib.parse.urlparse(self.directory + filename)
+
+    def has(self, filename):
+        raise NotImplementedError('use Index')
+
+    def write(self, filename, data):
+        headers = {
+            'Content-Type': mimetypes.guess_type(filename)[0] or 'text/plain',
+            s3.ACL: s3.PUBLIC
+        }
+        with s3.urlopen(self.url(filename), data=data, method='PUT', headers=headers) as response:
+            print(response.status, filename)
+
+    def delete(self, filenames):
+        # to do: multi-object delete API
+        for filename in filenames:
+            with s3.urlopen(self.url(filename), method='DELETE') as response:
+                print(response)
+
+
+class Index(Destination):
+    def __init__(self, destination, filename='index.html'):
+        self.destination = destination
+        self.filename = filename
+        self.files = set()
+        self.dirty = True
+
+    def has(self, filename):
+        return filename in self.files
+
+    def write(self, filename, data):
+        self.destination.write(filename, data)
+        self.files.add(filename)
+        self.dirty = True
+
+    def delete(self, filenames):
+        self.destination.delete(self.destination, filenames)
+        self.files.difference_update(filenames)
+        self.dirty = True
+
+    def sync(self):
+        if self.dirty:
+            self.destination.write(self.filename, textwrap.dedent('''
+                <html>
+                  <body>
+                    <h1>Directory listing for /</h1>
+                    <hr>
+                    <ul>''' + ''.join(f'''
+                      <li><a href={f}>{f}</a></li> ''' for f in self.files) + '''
+                    </ul>
+                  </body>
+                </html>
+                ''').encode('utf-8'))
+            self.dirty = False
+
+
+class AttachmentsDirectory:
+    def __init__(self, destination, local_directory):
+        self.destination = destination
+        self.path = local_directory
+
+    def scan(self):
+        for entry in os.scandir(self.path):
+            if not self.destination.has(entry.name) and entry.is_file(follow_symlinks=False):
+                with open(entry.path, 'rb') as file:
+                    data = file.read()
+                self.destination.write(entry.name, data)
+
+
+class ChunkedUploader:
+    SIZE_LIMIT = 1000000  # 1MB
+    TIME_LIMIT = 10       # 10s
```

Comment:
This was probably fine for testing, but for production one minute feels enough. Our logs are not that huge, so of these two it'll be the time limit which has the practical effect.

Reply:
Keep in mind that this doesn't mean that we send a block every 10 seconds.  It means that after we see some data, we wait for more data for 10 seconds before sending the accumulated data as a block.

I thought we could start low and increase it if we find that we're sending too much data...